#8702: fix some polymorphic variant error messages

Octachron opened the pull request:
This PR fixes a 4.08 regression reported in #8702 when mixing polymorphic variant and universal type variables
```OCaml
let foo : 'a . [`Foo] list -> ([> `Foo] as 'a) list =
    fun x -> x;;

Error: This expression has type [ `Foo ] list
       but an expression was expected of type [> `Foo ] list
       The second variant type does not allow tag(s) 
```
Along the way, it fixes a similarly confusing error message

```OCaml
type t = private [< `A]
let f (x:t): [> `A] =  x;;

Error: This expression has type t but an expression was expected of type
         [> `A ]
       Types for tag `A are incompatible
```
or

```OCaml
type any = Any: [< `A] -> any
let f (Any x): [ `A ] = x

Error: This expression has type [< `A ] but an expression was expected of type
         [ `A ]
       Types for tag `A are incompatible
```
Both confusing error messages stems from the fact that the error printer fails to notice that one of the row type involved was fixed (due to an universal type, a private type or a GADT).

This PR proposes to fix this issue by tracking in the types themselves the explanation for the fixed row. Then, the error message can print this explanation. For instance, the previous error messages become:

```OCaml
 Error: This expression has type [ `Foo ] list
       but an expression was expected of type [> `Foo ] list
       The second variant type is bound to the universal type variable 'a
```
```OCaml
Error: This expression has type t but an expression was expected of type
         [> `A ]
       The first variant type is private
```
```OCaml
Error: This expression has type [< `A ]
       but an expression was expected of type [ `A ]
       The first variant type is bound to $Any
```

Closes #8702

Octachron committed:
Fix mixed error messages and short explanation

trefis commented:
I'm going to say something that will sound absurd as soon as you put things back in context but: I'm not a fan of the proposed error messages.
The existing ones are of course awful, but I don't feel like the new ones are really adding much explanation.

How about:
```
# type t = private [< `A]
  let f (x:t): [> `A] =  x;;
Error: This expression has type t but an expression was expected of type
         [> `A ]
       The first variant type is private and may not allow tag(s) `A
```
which is still not great, but is at least no worse than:
```
# let f (x: [< `A]) : [> `A | `B] = x;;
Error: This expression has type [< `A ]
       but an expression was expected of type [> `A | `B ]
       The first variant type does not allow tag(s) `B
```

I guess the other messages you change can be updated in the same fashion.

---

As for the implementation itself: I see a lot of `row.row_fixed = None` and `row.row_fixed <> None`. I feel like an `is_fixed` predicate (working on rows) would be slightly nicer (or you could pattern match more often).

Octachron committed:
Review: improved error message

Octachron committed:
add Btype.is_fixed

Octachron committed:
update Changes

Octachron committed:
one more test

Octachron commented:
I agree that keeping the tag information is a good idea. I have split the error message in two subcases:
```OCaml
let f: 'a. ( [< `A] as 'a) -> [` A] = fun x -> x;;
```
> ```
>Error: This expression has type [< `A ]
>        but an expression was expected of type [ `A ]
>        The first variant type is bound to the universal type variable 'a,
>        it may not allow the tag `A 
>```
and
```OCaml
let f: 'a. [ `A | `B ] -> ([> `A ] as 'a) = fun x -> x
```
>```
>Error: This expression has type [ `A | `B ] but an expression was expected of type
>         [> `A ]
>       The second variant type is bound to the universal type variable 'a,
>       it cannot be closed
>```


I am wondering if the formulation is not too soft, and if we should replace `may not` with 

> it does not always allow the tag `A


On the implementation side,  I removed the comparisons to `None` with a `Btype.is_fixed` function .

trefis commented on typing/btype.ml:
```diff
@@ -427,16 +447,18 @@ let copy_row f fixed row keep more =
         | Reither(c, tl, m, e) ->
             let e = if keep then e else ref None in
-            let m = if row.row_fixed then fixed else m in
+            let m = if row.row_fixed <> None then fixed else m in
```
Shouldn't this be `if is_fixed row`?

trefis commented on typing/btype.mli:
```diff
@@ -69,8 +69,18 @@ val row_field: label -> row_desc -> row_field
+val fixed_explanation: row_desc -> fixed_explanation option
+(* Return the potential explanation for the fixed row *)
+val merge_fixed_explanation: row_desc -> row_desc -> fixed_explanation option
+(* Merge two explanations for a fixed row *)
```
Since you're already not respecting the same style as in the rest of this file (i.e. indenting the comments by like 36c), perhaps you could also add blank lines in between items?

:smiley:

Octachron replied:
You are right, let's add some breathing space.

trefis commented on typing/ctype.ml:
```diff
@@ -2800,12 +2809,13 @@ and unify_row env row1 row2 =
+    | None, Some _ -> rm2
+    | None, None -> newty2 (min rm1.level rm2.level) (Tvar None)
+  in
+  let fixed = merge_fixed_explanation row1 row2
```
That doesn't look right. Previously we were merging the result of the calls to `row_fixed`, here you're merging the `row_fixed` fields of `row1` and `row2`, these might be `None` while `fixed1` and `fixed2` might be `Some thing`.

Octachron replied:
Indeed, I have fixed the error by making `merge_fixed_explanation` type more symmetric as `fixed option -> fixed option -> fixed option`.

trefis commented on typing/ctype.ml:
```diff
@@ -2833,16 +2843,21 @@ and unify_row env row1 row2 =
   in
   let row0 = {row_fields = []; row_more = more; row_bound = ();
-              row_closed = closed; row_fixed = fixed; row_name = name} in
+              row_closed = closed; row_fixed=fixed; row_name = name} in
```
Why remove the space here?

trefis commented on typing/ctype.ml:
```diff
@@ -2833,16 +2843,21 @@ and unify_row env row1 row2 =
       else rest in
-    if rest <> [] && (row.row_closed || row_fixed row)
-    || closed && row_fixed row && not row.row_closed then begin
+    begin if rest <> [] && row.row_closed then
```
At first glance the disappearance of the `|| row_fixed row` (or equivalent) part seems like a mistake.
Is it really, or can you justify that it is actually correct?

Octachron replied:
You are right this was a major mistake. In particular, erroneous code like:
```OCaml
let f: 'a. ([> `A] as 'a) -> [> `A | `B | `C] = fun x -> x`
```
fell through the previous test.
The new code factorize the `row_fixed row` predicate without altering the full test.

trefis commented on typing/ctype.ml:
```diff
@@ -2879,13 +2894,19 @@ and unify_row env row1 row2 =
   | Reither(c1, tl1, m1, e1), Reither(c2, tl2, m2, e2) ->
       if e1 == e2 then () else
-      if (fixed1 || fixed2) && not (c1 || c2)
+      if (fixed1 <> None || fixed2 <> None) && not (c1 || c2)
```
Again with the `<> None`? :cry:

Octachron replied:
It is not here any more.

trefis commented on typing/ctype.ml:
```diff
@@ -2933,27 +2954,33 @@ and unify_row_field env fixed1 fixed2 more l f1 f2 =
+        update_scope rm.scope t1;
+        (try List.iter (unify env t1) tl
+         with exn -> e2 := None; raise exn)
+        )
```
Nitpick: the indentation is inconsistent with the previous case.

trefis commented on typing/printtyp.ml:
```diff
@@ -479,7 +479,7 @@ and raw_type_desc ppf = function
         "row_more=" raw_type row.row_more
         "row_closed=" row.row_closed
-        "row_fixed=" row.row_fixed
+        "row_fixed=" (Btype.is_fixed row)
```
That seems like a mistake. When I want to print the raw type, I want all the information in there.

At the very least, I'd want the information displayed to have the right type (so I would accept to see `Some _`, but not `true`).

Octachron replied:
I have upgraded the printer to print the full information.

Octachron committed:
fix handling of open fixed type errors

Octachron committed:
fix: merge the right fixity information

Octachron committed:
fix: full raw printer for fixed rows

Octachron committed:
fix indentation fix

Octachron committed:
fix: remove more "<> None"

trefis reviewed (approved):
LGTM.

trefis commented:
I'll let you do the actual merge; if you don't want to bother cleaning up the history before merging then please do squash everything.

Octachron commented:
I don't really see any intermediary point in the commit history worth to be saved. Would you prefer to have a trio of {base,reviewed,Change} commits rather than a squashed commit?

trefis commented:
Nope, squash is fine.

trefis commented:
This PR updates `swap_elt` to also handle variants, I think that part should be backported to 4.09 (this non-swapping is a regression in 4.08).

Octachron commented:
Good point, I will do so.
